oil_lang/expr_eval.py

```diff
@@ -1236,19 +1338,143 @@ def _EvalFuncCall(self, node):
     return ret
 
   def _EvalSubscript(self, node):
-    # type: (Subscript) -> Any # XXX
-    obj = self._EvalExpr(node.obj)
-    index = self._EvalIndices(node.indices)
-    try:
-      result = obj[index]
-    except KeyError:
-      # TODO: expr.Subscript has no error location
-      raise error.Expr('dict entry not found', loc.Missing)
-    except IndexError:
-      # TODO: expr.Subscript has no error location
-      raise error.Expr('index out of range', loc.Missing)
+    # type: (Subscript) -> value_t
+
+    obj = _PyObjToValue(self._EvalExpr(node.obj))
+    index = _PyObjToValue(self._EvalExpr(node.index))
+
+    UP_obj = obj
+    UP_index = index
+
+    with tagswitch(obj) as case:
+      if case(value_e.List):
+        obj = cast(value.List, UP_obj)
+        with tagswitch(index) as case2:
+          if case2(value_e.Slice):
+              index = cast(value.Slice, index)
+              step = 1
+              if index.step:
+                step = index.step.i
+
+              try:
+                  if index.lower and index.upper:
+                    return value.List(obj.items[index.lower.i:index.upper.i:step])
+
+                  elif index.lower:
+                    return value.List(obj.items[index.lower.i:len(obj.items):step])
+
+                  elif index.upper:
+                    return value.List(obj.items[:index.upper.i:step])
+
+                  else:
+                    # l[:] == l
+                    return value.List(list(obj.items))
+
+              except IndexError:
+                # TODO: expr.Subscript has no error location
+                raise error.Expr('index out of range', loc.Missing)
+
+          elif case2(value_e.Int):
+              index = cast(value.Int, index)
+              try:
+                return obj.items[index.i]
+              except IndexError:
+                # TODO: expr.Subscript has no error location
+                raise error.Expr('index out of range', loc.Missing)
+
+          else:
+            raise error.InvalidType('expected Slice or Int', loc.Missing)
+
+      elif case(value_e.Tuple):
```

Comment:
Language design note (I will copy these to Zulip) -- It's  a little weird we have Tuples, but we don't have tuples as `Dict` keys ...

I was wondering if maybe we use tuples mostly for object unpacking, like

```
return x, y

var x, y = Foo()

```

Tuples are kind of non-orthogonal in Python, I wonder if we can have named tuples , and immutable Dicts, etc.